### ISSUE TITLE
Get queue position of job

### ISSUE THREAD

**Issue by aparcar** [rq/rq#1197]
Hi, is it possible to get the queue position of a job? This would be useful to be available on enqueuing and also on checking the status, allowing the user to see some progress of requested jobs.

**Comment by selwin** [comment]
PR welcome :)

**Comment by aparcar** [comment]
Okay I will look into this, would be really handy for my use case

**Comment by aparcar** [comment]
@selwin Do you have some intuition on how to implement that? Checking the actual position in the build queue would be each time O(n) which seems a bit costly, instead it would be possible to store an increasing integer per job, which is then simply subtracted from the latest job integer ID. This would result in something like O(1) + O(1) (lookup latest integer ID + lookup job integer ID).
If there are multiple queues in priority, it could be `<calculated queue position> + <len(queues_with_higher_prio)>`.

I'll look a bit more into the current code base. Storing one extra `int` per job doesn't seem to harmful, however I'd be happy to hear your thoughts first.

**Comment by selwin** [comment]
I'm inclined to say let's just implement the method first so you can use it when you need to. The performance characteristics is similar to `queue.remove(job)` which is also `O(N)`.

**Comment by aparcar** [comment]
@selwin I haven't looked deeper in the `rq` codebase yet but sounds like an interesting thing to do. If you're up for the task I can also happily pass it on...

**Comment by davidmurray** [comment]
This would be quite useful for me too. Any update on this ?

Thanks,
David

**Comment by aparcar** [comment]
@selwin so `queue.remove()` uses LREM which does exactly what we want, except it removes also the entry. We would need a Redis function like `IRANK` to return the position without removing the element as well. I'll look into that, however doubt that upstream Redis will merge this really.

Second plan is to use HSET and give every job a *position ID*.

**Comment by davidmurray** [comment]
In case anyone is looking for a naive implementation for the time being, here is my implementation:
```python

if job.id in queue.job_ids:
        position_in_queue = queue.job_ids.index(job.id)

```

**Comment by selwin** [comment]
@aparcar I think we could implement `job.get_position()` like @davidmurray suggested for now. However, we should document what this command does and warn that it may cause poor performance on queues that are super large.

After `redis.lpos()` is available, we can always switch the implementation to a more efficient one (even though even inside Redis it's still O(N)).

**Comment by aparcar** [comment]
@selwin should this be Job().get_position or rahter Queue().get_job_position(job_id)?

**Comment by selwin** [comment]
I think both should be possible. I’d think job.get_position() calls queue.get_job_position() under the hood.

What do you think?

**Comment by aparcar** [comment]
Fine for me, I'm wondering if the performance is even worse if we import Queue every time like `cancel()` does https://github.com/rq/rq/blob/master/rq/job.py#L571

**Comment by aparcar** [comment]
Implemented via #1271. It's *ugly* in a way that both `if id in job_ids` and `job_ids.index(job_id)` have a worst case complexity of O(N). However it's later trivially replaced with `redis.lpos()`

**Comment by davidmurray** [comment]
> Fine for me, I'm wondering if the performance is even worse if we import Queue every time like `cancel()` does https://github.com/rq/rq/blob/master/rq/job.py#L571

As far as I know, if the module is imported once then subsequent imports should take no time.  This should not be a problem. (see https://stackoverflow.com/a/19077396)
